### PR TITLE
Fix `yield_sequences_from_token_batch` to avoid losing the removed token when `BOS` is prepended

### PR DESCRIPTION
**Pull request by jukofyork**
I've just been experimenting with setting `batch_size_tokens` >> `sequence_len` and noticed one token was getting lost for each `BOS` insertion.

I think this is the correct solution and prepends it back to the `tokens` list for reuse? It's possibly not very efficient but not sure it makes much difference due to training taking so long comparably?

### CODE REVIEW

**Comment by jukofyork** [comment]
Can confirm this is working, but I'm yet unsure if using `batch_size_tokens != sequence_len` is a horrible idea (it seems to get a similar loss to for the equivalent `batch_size_tokens = sequence_len` for smaller `sequence_len` for `command-r` models at least.

**Comment by jukofyork** [comment]
So I've been reading up more on this today and it seems that the current implementation won't work well then `batch_size_tokens` > `sequence_len` due to having multiple `<BOS>` tokens.

It looks like the best available way to handle this would be:

```
<BOS> sequence 1 text <EOS> sequence 2 text <EOS> sequence 3 text...
```

It seems important that their be just a single `<BOS>` like this at `position_id = 0` as the models are hypothesised to use it as a "sink" to balance out Softmax probabilities having to sum to 1, etc. The people working on `llama.cpp` seem to have confirmed this and their `imatrix` generation code just adds this to the start of each "chunk" and takes no notice of anything else regarding special tokens or chat templates...

For some use cases, we might also want to set all `<EOS>` tokens to have the `labels` value to `-100` (to not use it for the loss and avoid biasing the response lengths, etc).

There is no easy way to do proper sample packing as the transformers code is all setup to use a 1D `attention_mask` (which is only really useful for skipping `<PAD>` tokens):

https://github.com/huggingface/transformers/issues/25452 

There is this code which patches some models to work with sample packing:

https://github.com/MeetKai/functionary/tree/main/functionary/train/packing

but I don't think this is really an option, and looks very brittle and unmaintained.

I'll leave this PR up as it does save the extra token that was getting written over using the `<BOS>` token, but it may be worth printing a warning if `batch_size_tokens` > `sequence_len`.

---

If I get time, then I will look into adding a new dataset format that 

1. Uses the best available method of:

```
<BOS> sequence 1 text <EOS> sequence 2 text <EOS> sequence 3 text...
```

2. Allows `<EOS>` labels to be set to `-100`.

and possibly:

- Add the option to split the sequences at better points (eg: whitespace, punctuation or newlines).
- Add the option to "mask" some parts for the loss by setting the labels to `-100`.

as these might then let use teach models things like authors' style by adding some masked contextual information like this:

```
<BOS> # Cormac McCarthy, Blood Meridian, 1985

It makes no difference what men think of war, said the judge. War endures. As well ask men what they think of stone. War was always here. Before man was, war waited for him. The ultimate trade awaiting its ultimate practitioner. That is the way it was and will be. That way and not some other way.
<EOS> # Patricia Highsmith, The Price of Salt, 1952

January. It was all things. And it was one thing, like a solid door. Its cold sealed the city in a gray capsule. January was moments, and January was a year. January rained the moments down, and froze them in her memory: [...]Every human action seemed to yield a magic. January was a two-faced month, jangling like jester's bells, crackling like snow crust, pure as any beginning, grim as an old man, mysteriously familiar yet unknown, like a word one can almost but not quite define.
<EOS> # ...
```

or similar, with the "# Cormac McCarthy, Blood Meridian, 1985" and `<EOS>` tokens having their labels all set to `-100`.

**Comment by jukofyork** [comment]
I'm closing this as I think the `yield_sequences_from_token_batch` code likely wants a more through refactoring at a later date and this small fix isn't really worthwhile in isolation.

---

I've managed to get a version that takes files with the `<EOS>` tokens inside the text files like this:

```
Not unkindly, Mr. Nell told him, "There's two parts to the system. One part carries solid human waste--shit, if I'd not be offendin yer tender ears. The other part carries gray water--water flushed from toilets or run down the drains from sinks and washin-machines and showers; it's also the water that runs down the gutters into the city drains.

<|END_OF_TURN_TOKEN|>The aluminum sled on which Norah was transporting her testing gear resembled an oversized Flexible Flyer. The craft was prepacked with diagnostic gear and safety accessories she'd been using on the glacier over the past few days. All of her gear--including a battery pack, safety flares, and a powerful front-mounted spotlight--was bound under a secured, plastic tarp. Despite the heavy load, the sled glided effortlessly on long, straight runners. Even on the almost imperceptible incline, the sled moved downhill on its own accord, and Norah applied a gentle restraint, almost as if allowing the sled to lead the way. Sensing the distance growing between the group and the habisphere, Tolland looked over his shoulder. Only fifty yards away, the pale curvature of the dome had all but disappeared in the blustery blackness.

<|END_OF_TURN_TOKEN|>He packed a compartmentalized, hand-tooled Mark Cross briefcase with the blue bag, the Green Acres bag, and the tape recorder that he used for dictation. While he waited for the Keanuphobe to call, he would do some game planning and compose a chapter *of Fear Not for l Am with You.*

<|END_OF_TURN_TOKEN|>Well, the word was out. Cancer. Rhymes with *dancer* and You *just shit your pants, sir.* God knew the word had bobbed up in his own mind more than once since getting on the penny scale in front of the shoe store. It had bobbed up like some evil clown's dirty balloon and he had turned away from it. He had turned away from it the way you turned away from the bag ladies who sat rocking back and forth in their strange, sooty little nooks outside the Grand Central Station or the way you turned away from the capering Gypsy children who had come with the rest of the Gypsy band. The Gypsy children sang in voices that somehow managed to be both monotonous and strangely sweet at the same time. The Gypsy children walked on their hands with tambourines outstretched, held somehow by their bare dirty toes. The Gypsy children juggled. The Gypsy children put the local Frisbee jocks to shame by spinning two, sometimes three of the plastic disks at the same time - on fingers, on thumbs, sometimes on noses. They laughed while they did all those things, and they all seemed to have skin diseases or crossed eyes or harelips. When you suddenly found such a weird combination of agility and ugliness thrust in front of you, what else was there to do but turn away? Bag ladies, Gypsy children, and cancer. Even the skittery run of his thoughts frightened him.

<|END_OF_TURN_TOKEN|>
```

but all the work of extracting, shuffling, formatting, etc is done using bash scripts as python was so slow it kept timing out the Deepspeed connection...

```python
def yield_sequences_from_token_batch(tokenizer, token_batch, sequence_len):
    """Yields fixed-length sequences from batches of tokens, ensuring proper BOS/EOS token handling.
    
    Takes batches of tokens and yields sequences of fixed length, with each sequence:
    - Starting with BOS token if specified in tokeniser
    - Containing complete chunks terminated by EOS tokens (never splitting between EOS tokens)
    - Right-padded with extra EOS tokens if needed so all reach exactly sequence_len
    """
    sequence_tokens = [] if tokenizer.bos_token_id is None else [tokenizer.bos_token_id]
    for tokens in tqdm(token_batch):
        tokens = tokens.tolist()
        assert len(tokens) > 0, "empty token list"
        assert tokens[-1] == tokenizer.eos_token_id, "token lists must end with EOS"

        idx = 0
        # If present, skip the auto-generated BOS token
        if tokenizer.bos_token_id is not None and tokens[0] == tokenizer.bos_token_id:
            idx += 1

        while idx < len(tokens):          
            next_eos_idx = tokens.index(tokenizer.eos_token_id, idx)
            chunk = tokens[idx:next_eos_idx + 1]
            assert len(chunk) <= sequence_len, "chunk exceeds sequence length"
 
            if len(sequence_tokens) + len(chunk) > sequence_len:
                sequence_tokens.extend([tokenizer.eos_token_id] * (sequence_len - len(sequence_tokens)))
                yield sequence_tokens
                sequence_tokens = [] if tokenizer.bos_token_id is None else [tokenizer.bos_token_id]

            sequence_tokens.extend(chunk)
            idx += len(chunk)

    if len(sequence_tokens) > 0:
        sequence_tokens.extend([tokenizer.eos_token_id] * (sequence_len - len(sequence_tokens)))
        yield sequence_tokens
```

This then gets called like this:

```python
    dataset = dataset.map(lambda x: tokenizer(x['text']), batched=True, batch_size=10, remove_columns=dataset.column_names, desc='tokenizing', num_proc=num_proc)
    dataset = dataset.map(lambda x: {'input_ids': list(yield_sequences_from_token_batch(tokenizer, x['input_ids'], sequence_len))}, batched=True, batch_size=None, remove_columns=dataset.column_names, desc='splitting')
    # Set labels for EOS tokens -100 to exclude them from training gradient calculations
    dataset = dataset.map(
        lambda x: {
            'attention_mask': torch.ones_like(x['input_ids']),
            'labels': torch.where(x['input_ids'] == tokenizer.eos_token_id, torch.full_like(x['input_ids'], -100), x['input_ids'])
        },
        desc='adding attention_mask and labels (with EOS labels set to -100)'
    )
```

To ensure the `<EOS>` tokens are attended to, but not used for gradient calculations (which means I can right-pad all the batches up to the desired sequence length).

This is likely not useful for others though and very specific for what I am trying to do, but might be interesting or useful for a base to refactor `yield_sequences_from_token_batch` for more general use later.
